### PR TITLE
refactor(melange): separate rule loading

### PR DESCRIPTION
**Pull request by rgrinberg**
.cmj and .js rules should be separated

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

ps-id: 6362ec95-ee71-4c56-97ac-8144b24a28e5

### CODE REVIEW

File: test/blackbox-tests/test-cases/melange/empty-entries-target-dir-exists.t [hunk 1]
```diff
@@ -17,5 +17,9 @@ Create the dist folder
   $ mkdir ./dist
 
   $ dune build @melange-dist
+  Error: No rule found for dist/melange.js
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/empty-entries-target-dir-exists.t:20]: @jchavarri i don't understand how this test worked before. Where does the `melange.js` file come from?

Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/empty-entries-target-dir-exists.t:20]: I don't know either. @anmonteiro added the test in https://github.com/ocaml/dune/pull/6506 and then modified it in https://github.com/ocaml/dune/pull/6579/.

File: src/dune_rules/buildable_rules.mli [hunk 2]
```diff
@@ -31,6 +31,18 @@ type kind =
       ; empty_module_interface_if_absent : bool
       }
 
+(** temporary hacks until we remove preprocssing settings from the cctx *)
```
Comment (jchavarri) [inline src/dune_rules/buildable_rules.mli:34]:
```suggestion
(** temporary hacks until we remove preprocessing settings from the cctx *)
```

File: src/dune_rules/buildable_rules.mli [hunk 3]
```diff
@@ -31,6 +31,18 @@ type kind =
...
+val pp_spec :
```
Comment (jchavarri) [inline src/dune_rules/buildable_rules.mli:35]: I don't understand why this function has to be exposed, I don't see it used outside `Buildable_rules`.

File: test/blackbox-tests/test-cases/melange/empty-entries-target-dir-exists.t (hunk 1 above)
Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/empty-entries-target-dir-exists.t:20]: I’m not sure where it comes from either. I assumed it was the empty wrapper when entries are empty

File: src/dune_rules/gen_rules.ml [hunk 4]
```diff
@@ -438,10 +420,36 @@ let gen_rules ~sctx ~dir components : Build_config.gen_rules_result Memo.t =
...
+          let parent = Path.Build.parent_exn dir in
+          let+ stanzas = Only_packages.stanzas_in_dir parent in
+          match
+            match stanzas with
+            | None -> None
+            | Some stanzas ->
+              List.find_map stanzas.stanzas ~f:(function
+                | Melange_emit mel ->
+                  let target_dir = Path.Build.relative parent mel.target in
+                  Option.some_if (Path.Build.equal target_dir dir) mel
```
Comment (jchavarri) [inline src/dune_rules/gen_rules.ml:436]: I understand the purpose of the new code is to generate js rules from the directory where the melange.emit stanza lives (right?). But what is the code in these two lines doing specifically? Why do we compare `target_dir` with `dir`?

**Review by jchavarri — APPROVED** [review]

File: src/dune_rules/gen_rules.ml (hunk 4 above)
Comment (rgrinberg) [reply src/dune_rules/gen_rules.ml:436]: > I understand the purpose of the new code is to generate js rules from the directory where the melange.emit stanza lives (right?). 

Actually, it is to generate the js rules from target directory of the melange stanza. So if we have `(melange.emit (target foo))` in `src/`, we'd like to generate the rules in `src/foo`.

> But what is the code in these two lines doing specifically? Why do we compare target_dir with dir?

This is where we detect the case above.

File: src/dune_rules/buildable_rules.mli (hunk 3 above)
Comment (rgrinberg) [reply src/dune_rules/buildable_rules.mli]: Good catch, it must have been in some older iteration. I got rid of it.

File: test/blackbox-tests/test-cases/melange/empty-entries-target-dir-exists.t (hunk 1 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/melange/empty-entries-target-dir-exists.t]: > I assumed it was the empty wrapper when entries are empty

That's indeed the reason. Perhaps we should give a more obvious name.

Anyway this caught a bug which I now fixed.

**Pull request merged**